rust/lance-encoding/src/encodings/physical/bitpack_fastlanes.rs

```diff
@@ -0,0 +1,1547 @@
+
+        DataType::UInt32 => {
+            let mut global_max: u32 = 0;
+            for array in arrays {
+                let primitive_array = array
+                    .as_any()
+                    .downcast_ref::<PrimitiveArray<UInt32Type>>()
+                    .unwrap();
+                let array_max = arrow::compute::bit_or(primitive_array).unwrap_or(0);
+                global_max = global_max.max(array_max);
+            }
+            let num_bits =
+                arrays[0].data_type().byte_width() as u64 * 8 - global_max.leading_zeros() as u64;
+            if num_bits == 0 {
+                res = 1;
+            } else {
+                res = num_bits;
+            }
+        }
+
+        DataType::Int32 => {
+            let mut global_max_width: u64 = 0;
+            for array in arrays {
+                let primitive_array = array
+                    .as_any()
+                    .downcast_ref::<PrimitiveArray<Int32Type>>()
+                    .unwrap();
+                let array_max_width = arrow::compute::bit_or(primitive_array).unwrap_or(0);
+                global_max_width =
+                    global_max_width.max(32 - array_max_width.leading_zeros() as u64);
+            }
+            if global_max_width == 0 {
+                res = 1;
+            } else {
+                res = global_max_width;
+            }
+        }
+
+        DataType::UInt64 => {
+            let mut global_max: u64 = 0;
+            for array in arrays {
+                let primitive_array = array
+                    .as_any()
+                    .downcast_ref::<PrimitiveArray<UInt64Type>>()
+                    .unwrap();
+                let array_max = arrow::compute::bit_or(primitive_array).unwrap_or(0);
+                global_max = global_max.max(array_max);
+            }
+            let num_bits =
+                arrays[0].data_type().byte_width() as u64 * 8 - global_max.leading_zeros() as u64;
+            if num_bits == 0 {
+                res = 1;
+            } else {
+                res = num_bits;
+            }
+        }
+
+        DataType::Int64 => {
+            let mut global_max_width: u64 = 0;
+            for array in arrays {
+                let primitive_array = array
+                    .as_any()
+                    .downcast_ref::<PrimitiveArray<Int64Type>>()
+                    .unwrap();
+                let array_max_width = arrow::compute::bit_or(primitive_array).unwrap_or(0);
+                global_max_width =
+                    global_max_width.max(64 - array_max_width.leading_zeros() as u64);
+            }
+            if global_max_width == 0 {
+                res = 1;
+            } else {
+                res = global_max_width;
+            }
+        }
+        _ => {
+            panic!("BitpackedForNonNegArrayEncoder only supports data types of UInt8, Int8, UInt16, Int16, UInt32, Int32, UInt64, Int64");
+        }
+    };
+    res
+}
+
+macro_rules! encode_fixed_width {
```

Comment:
Please document these macros.  Rust macros aren't as evil as C macros but they can still be hard to read.  What is this macro expected to do?

Reply:
fixed